File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.py
n = int(input())
li = list(map(int, input().split()))
dp = [[1]*2 for _ in range(n)]
for i in range(n):
    for j in range(i):
        if li[i] > li[j]:
            dp[i][0] = max(dp[i][0], dp[j][0]+1)
for i in range(n-1, -1, -1):
    for j in range(n-1, i, -1):
        if li[i] > li[j]:
            dp[i][1] = max(dp[i][1], dp[j][1]+1)
res = 0    
for i in range(n):
    cur = dp[i][0] + dp[i][1] - 1
    if cur > res:
        res = cur
print(res)

File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] dp1 = new int[n];
        int[] dp2 = new int[n];
        int[] arr = new int[n];
        Arrays.fill(dp1, 1);
        Arrays.fill(dp2, 1);
        for (int i = 0; i < n ;i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] > arr[j]) {
                    dp1[i] = Math.max(dp1[i], dp1[j] + 1);
                }
            }
        }

        for (int i = n-2; i >= 0; i--) {
            for (int j = n-1; j > i; j--) {
                if (arr[i] > arr[j]) {
                    dp2[i] = Math.max(dp2[i], dp2[j] + 1);
                }
            }
        }

        int res = 0;
        for (int i = 0; i < n; i++) {
            res = Math.max(res, dp1[i] + dp2[i] - 1);
        }
        System.out.println(res);
    }
}
